In [1]:
from pydataset import data

In [2]:
import pandas as pd
import numpy as np

np.random.seed(123)

students = ['Sally', 'Jane', 'Suzie', 'Billy', 'Ada', 'John', 'Thomas',
            'Marie', 'Albert', 'Richard', 'Isaac', 'Alan']

# randomly generate scores for each student for each subject
# note that all the values need to have the same length here
math_grades = np.random.randint(low=60, high=100, size=len(students))
english_grades = np.random.randint(low=60, high=100, size=len(students))
reading_grades = np.random.randint(low=60, high=100, size=len(students))

df = pd.DataFrame({'name': students,
                   'math': math_grades,
                   'english': english_grades,
                   'reading': reading_grades})

type(df)


pandas.core.frame.DataFrame

In [3]:
df

name  math  english  reading
0     Sally    62       85       80
1      Jane    88       79       67
2     Suzie    94       74       95
3     Billy    98       96       88
4       Ada    77       92       98
5      John    79       76       93
6    Thomas    82       64       81
7     Marie    93       63       90
8    Albert    92       62       87
9   Richard    69       80       94
10    Isaac    92       99       93
11     Alan    92       62       72

Create a column named passing_english that indicates whether each student has a passing grade in english.

In [4]:
df["passing_english"] = df.english >= 70
df

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
11     Alan    92       62       72            False

In [5]:
df.sort_values(by='passing_english', ascending=False)

name  math  english  reading  passing_english
0     Sally    62       85       80             True
1      Jane    88       79       67             True
2     Suzie    94       74       95             True
3     Billy    98       96       88             True
4       Ada    77       92       98             True
5      John    79       76       93             True
9   Richard    69       80       94             True
10    Isaac    92       99       93             True
6    Thomas    82       64       81            False
7     Marie    93       63       90            False
8    Albert    92       62       87            False
11     Alan    92       62       72            False

Sort the english grades first by passing_english and then by student name. 

All the students that are failing english should be first, and within the students that are failing english they should be ordered alphabetically.

 The same should be true for the students passing english. (Hint: you can pass a list to the .sort_values method)

In [6]:
df.sort_values(by=['passing_english','name'])

name  math  english  reading  passing_english
11     Alan    92       62       72            False
8    Albert    92       62       87            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True
1      Jane    88       79       67             True
5      John    79       76       93             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
2     Suzie    94       74       95             True

Sort the english grades first by passing_english, and then by the actual english grade, similar to how we did in the last step.

In [7]:
df.sort_values(by=['passing_english','english'])

name  math  english  reading  passing_english
8    Albert    92       62       87            False
11     Alan    92       62       72            False
7     Marie    93       63       90            False
6    Thomas    82       64       81            False
2     Suzie    94       74       95             True
5      John    79       76       93             True
1      Jane    88       79       67             True
9   Richard    69       80       94             True
0     Sally    62       85       80             True
4       Ada    77       92       98             True
3     Billy    98       96       88             True
10    Isaac    92       99       93             True

Calculate each students overall grade and add it as a column on the dataframe. The overall grade is the average of the math, english, and reading grades.

In [8]:
df["average_grade"] = (df.math + df.english + df.reading) / 3
df

name  math  english  reading  passing_english  average_grade
0     Sally    62       85       80             True      75.666667
1      Jane    88       79       67             True      78.000000
2     Suzie    94       74       95             True      87.666667
3     Billy    98       96       88             True      94.000000
4       Ada    77       92       98             True      89.000000
5      John    79       76       93             True      82.666667
6    Thomas    82       64       81            False      75.666667
7     Marie    93       63       90            False      82.000000
8    Albert    92       62       87            False      80.333333
9   Richard    69       80       94             True      81.000000
10    Isaac    92       99       93             True      94.666667
11     Alan    92       62       72            False      75.333333

2 Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:

In [9]:
# data('mpg', show_doc=True) # view the documentation for the dataset
mpg = data('mpg') # load the dataset and store it in a variable

In [10]:
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [11]:
len(mpg)

234

In [12]:
mpg.size

2574

In [13]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displ         234 non-null    float64
 3   year          234 non-null    int64  
 4   cyl           234 non-null    int64  
 5   trans         234 non-null    object 
 6   drv           234 non-null    object 
 7   cty           234 non-null    int64  
 8   hwy           234 non-null    int64  
 9   fl            234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


## Load the mpg dataset. Read the documentation for the dataset and use it for the following questions:
- How many rows and columns are there?
There are 234 rows and 11 columns
- What are the data types of each column?
THE DATA TYPE INFO CAN BE FOUND IN THE GRAPH ABOVE
- Summarize the dataframe with .info and .describe
DONE ABOVE
- Rename the cty column to city.
DONE
- Rename the hwy column to highway.
DONE
- Do any cars have better city mileage than highway mileage?
NO
- Create a column named mileage_difference this column should contain the difference between highway and city mileage for each car.
DONE
- Which car (or cars) has the highest mileage difference?
DONE
- Which compact class car has the lowest highway mileage? The best?

- Create a column named average_mileage that is the mean of the city and highway mileage.
- Which dodge car has the best average mileage? The worst?

In [14]:
new_name = {
    "cty": "city",
    "hwy": "highway"
}
new_name

{'cty': 'city', 'hwy': 'highway'}

In [15]:
mpg = mpg.rename(columns=new_name)

In [16]:
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

In [17]:
mpg[(mpg.city > mpg.highway)]

Empty DataFrame
Columns: [manufacturer, model, displ, year, cyl, trans, drv, city, highway, fl, class]
Index: []

In [18]:
mpg["mileage_difference"] = mpg.highway - mpg.city
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

In [19]:
mpg["mileage_difference"]

1      11
2       8
3      11
4       9
5      10
       ..
230     9
231     8
232    10
233     8
234     9
Name: mileage_difference, Length: 234, dtype: int64

In [20]:
mpg.sort_values(by="mileage_difference", ascending=False)

manufacturer              model  displ  year  cyl       trans drv  city  \
107        honda              civic    1.8  2008    4    auto(l5)   f    24   
223   volkswagen         new beetle    1.9  1999    4    auto(l4)   f    29   
1           audi                 a4    1.8  1999    4    auto(l5)   f    18   
229   volkswagen             passat    1.8  1999    4    auto(l5)   f    18   
36     chevrolet             malibu    3.5  2008    6    auto(l4)   f    18   
..           ...                ...    ...   ...  ...         ...  ..   ...   
80          ford       explorer 4wd    4.0  1999    6    auto(l5)   4    14   
138      mercury    mountaineer 4wd    4.0  1999    6    auto(l5)   4    14   
177       toyota        4runner 4wd    3.4  1999    6  manual(m5)   4    15   
152       nissan     pathfinder 4wd    3.3  1999    6  manual(m5)   4    15   
204       toyota  toyota tacoma 4wd    3.4  1999    6  manual(m5)   4    15   

     highway fl       class  mileage_difference  
107       36  c  subcompact                  12  
223       41  d  subcompact                  12  
1         29  p     compact                  11  
229       29  p     midsize                  11  
36        29  r     midsize                  11  
..       ... ..         ...                 ...  
80        17  r         suv                   3  
138       17  r         suv                   3  
177       17  r         suv                   2  
152       17  r         suv                   2  
204       17  r      pickup                   2  

[234 rows x 12 columns]

In [21]:
# mpg.sort_values(by=["highway"])#.mpg.class.str.startswith("c")

In [22]:
mpg[mpg["class"] == "compact"].sort_values(by="highway", ascending=False)

manufacturer         model  displ  year  cyl       trans drv  city  \
213   volkswagen         jetta    1.9  1999    4  manual(m5)   f    33   
197       toyota       corolla    1.8  2008    4  manual(m5)   f    28   
196       toyota       corolla    1.8  1999    4  manual(m5)   f    26   
198       toyota       corolla    1.8  2008    4    auto(l4)   f    26   
195       toyota       corolla    1.8  1999    4    auto(l4)   f    24   
189       toyota  camry solara    2.4  2008    4  manual(m5)   f    21   
3           audi            a4    2.0  2008    4  manual(m6)   f    20   
190       toyota  camry solara    2.4  2008    4    auto(s5)   f    22   
4           audi            a4    2.0  2008    4    auto(av)   f    21   
194       toyota       corolla    1.8  1999    4    auto(l3)   f    24   
208   volkswagen           gti    2.0  1999    4  manual(m5)   f    21   
211   volkswagen           gti    2.0  2008    4    auto(s6)   f    22   
210   volkswagen           gti    2.0  2008    4  manual(m6)   f    21   
188       toyota  camry solara    2.2  1999    4  manual(m5)   f    21   
214   volkswagen         jetta    2.0  1999    4  manual(m5)   f    21   
216   volkswagen         jetta    2.0  2008    4    auto(s6)   f    22   
217   volkswagen         jetta    2.0  2008    4  manual(m6)   f    21   
218   volkswagen         jetta    2.5  2008    5    auto(s6)   f    21   
219   volkswagen         jetta    2.5  2008    5  manual(m5)   f    21   
2           audi            a4    1.8  1999    4  manual(m5)   f    21   
1           audi            a4    1.8  1999    4    auto(l5)   f    18   
142       nissan        altima    2.4  1999    4  manual(m5)   f    21   
10          audi    a4 quattro    2.0  2008    4  manual(m6)   4    20   
187       toyota  camry solara    2.2  1999    4    auto(l4)   f    21   
173       subaru   impreza awd    2.5  2008    4  manual(m5)   4    20   
171       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
7           audi            a4    3.1  2008    6    auto(av)   f    18   
193       toyota  camry solara    3.3  2008    6    auto(s5)   f    18   
143       nissan        altima    2.4  1999    4    auto(l4)   f    19   
11          audi    a4 quattro    2.0  2008    4    auto(s6)   4    19   
5           audi            a4    2.8  1999    6    auto(l5)   f    16   
6           audi            a4    2.8  1999    6  manual(m5)   f    18   
8           audi    a4 quattro    1.8  1999    4  manual(m5)   4    18   
215   volkswagen         jetta    2.0  1999    4    auto(l4)   f    19   
209   volkswagen           gti    2.0  1999    4    auto(l4)   f    19   
192       toyota  camry solara    3.0  1999    6  manual(m5)   f    18   
191       toyota  camry solara    3.0  1999    6    auto(l4)   f    18   
12          audi    a4 quattro    2.8  1999    6    auto(l5)   4    15   
13          audi    a4 quattro    2.8  1999    6  manual(m5)   4    17   
14          audi    a4 quattro    3.1  2008    6    auto(s6)   4    17   
9           audi    a4 quattro    1.8  1999    4    auto(l5)   4    16   
15          audi    a4 quattro    3.1  2008    6  manual(m6)   4    15   
170       subaru   impreza awd    2.5  2008    4    auto(s4)   4    20   
172       subaru   impreza awd    2.5  2008    4  manual(m5)   4    19   
212   volkswagen           gti    2.8  1999    6  manual(m5)   f    17   
221   volkswagen         jetta    2.8  1999    6  manual(m5)   f    17   
220   volkswagen         jetta    2.8  1999    6    auto(l4)   f    16   

     highway fl    class  mileage_difference  
213       44  d  compact                  11  
197       37  r  compact                   9  
196       35  r  compact                   9  
198       35  r  compact                   9  
195       33  r  compact                   9  
189       31  r  compact                  10  
3         31  p  compact                  11  
190       31  r  compact                   9  
4         30  p  compact    

In [23]:
mpg.head()

manufacturer model  displ  year  cyl       trans drv  city  highway fl  \
1         audi    a4    1.8  1999    4    auto(l5)   f    18       29  p   
2         audi    a4    1.8  1999    4  manual(m5)   f    21       29  p   
3         audi    a4    2.0  2008    4  manual(m6)   f    20       31  p   
4         audi    a4    2.0  2008    4    auto(av)   f    21       30  p   
5         audi    a4    2.8  1999    6    auto(l5)   f    16       26  p   

     class  mileage_difference  
1  compact                  11  
2  compact                   8  
3  compact                  11  
4  compact                   9  
5  compact                  10

In [24]:
mpg["average_mileage"] = (mpg.highway +  mpg.city) / 2
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

In [25]:
mpg.sort_values(by="average_mileage", ascending=False)

manufacturer                model  displ  year  cyl       trans drv  city  \
222   volkswagen           new beetle    1.9  1999    4  manual(m5)   f    35   
213   volkswagen                jetta    1.9  1999    4  manual(m5)   f    33   
223   volkswagen           new beetle    1.9  1999    4    auto(l4)   f    29   
197       toyota              corolla    1.8  2008    4  manual(m5)   f    28   
196       toyota              corolla    1.8  1999    4  manual(m5)   f    26   
..           ...                  ...    ...   ...  ...         ...  ..   ...   
127         jeep   grand cherokee 4wd    4.7  2008    8    auto(l5)   4     9   
60         dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
66         dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70         dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   
55         dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   

     highway fl       class  mileage_difference  average_mileage  
222       44  d  subcompact                   9             39.5  
213       44  d     compact                  11             38.5  
223       41  d  subcompact                  12             35.0  
197       37  r     compact                   9             32.5  
196       35  r     compact                   9             30.5  
..       ... ..         ...                 ...              ...  
127       12  e         suv                   3             10.5  
60        12  e         suv                   3             10.5  
66        12  e      pickup                   3             10.5  
70        12  e      pickup                   3             10.5  
55        12  e      pickup                   3             10.5  

[234 rows x 13 columns]

In [30]:
mpg[mpg.manufacturer.str.startswith("dodge")].sort_values(by="average_mileage",ascending=False)

manufacturer                model  displ  year  cyl       trans drv  city  \
38        dodge          caravan 2wd    2.4  1999    4    auto(l3)   f    18   
42        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
43        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    17   
39        dodge          caravan 2wd    3.0  1999    6    auto(l4)   f    17   
47        dodge          caravan 2wd    3.8  2008    6    auto(l6)   f    16   
48        dodge          caravan 2wd    4.0  2008    6    auto(l6)   f    16   
40        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
41        dodge          caravan 2wd    3.3  1999    6    auto(l4)   f    16   
45        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
46        dodge          caravan 2wd    3.8  1999    6    auto(l4)   f    15   
49        dodge    dakota pickup 4wd    3.7  2008    6  manual(m6)   4    15   
53        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
54        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4    14   
50        dodge    dakota pickup 4wd    3.7  2008    6    auto(l4)   4    14   
52        dodge    dakota pickup 4wd    3.9  1999    6  manual(m5)   4    14   
63        dodge          durango 4wd    5.7  2008    8    auto(l5)   4    13   
51        dodge    dakota pickup 4wd    3.9  1999    6    auto(l4)   4    13   
73        dodge  ram 1500 pickup 4wd    5.7  2008    8    auto(l5)   4    13   
68        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
58        dodge          durango 4wd    3.9  1999    6    auto(l4)   4    13   
59        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
67        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4    13   
61        dodge          durango 4wd    4.7  2008    8    auto(l5)   4    13   
69        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
65        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4    12   
56        dodge    dakota pickup 4wd    5.2  1999    8  manual(m5)   4    11   
44        dodge          caravan 2wd    3.3  2008    6    auto(l4)   f    11   
62        dodge          durango 4wd    5.2  1999    8    auto(l4)   4    11   
72        dodge  ram 1500 pickup 4wd    5.2  1999    8  manual(m5)   4    11   
64        dodge          durango 4wd    5.9  1999    8    auto(l4)   4    11   
57        dodge    dakota pickup 4wd    5.2  1999    8    auto(l4)   4    11   
71        dodge  ram 1500 pickup 4wd    5.2  1999    8    auto(l4)   4    11   
74        dodge  ram 1500 pickup 4wd    5.9  1999    8    auto(l4)   4    11   
66        dodge  ram 1500 pickup 4wd    4.7  2008    8    auto(l5)   4     9   
60        dodge          durango 4wd    4.7  2008    8    auto(l5)   4     9   
55        dodge    dakota pickup 4wd    4.7  2008    8    auto(l5)   4     9   
70        dodge  ram 1500 pickup 4wd    4.7  2008    8  manual(m6)   4     9   

    highway fl    class  mileage_difference  average_mileage  
38       24  r  minivan                   6             21.0  
42       24  r  minivan                   7             20.5  
43       24  r  minivan                   7             20.5  
39       24  r  minivan                   7             20.5  
47       23  r  minivan                   7             19.5  
48       23  r  minivan                   7             19.5  
40       22  r  minivan                   6             19.0  
41       22  r  minivan                   6             19.0  
45       22  r  minivan                   7             18.5  
46       21  r  minivan                   6             18.0  
49       19  r   pickup                   4             17.0  
53       19  r   pickup                   5             16.5  
54       19  r   pickup                   5             16.5  
50       18  r   pickup                   4             16.0  
52       17  r   